File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words;
        string word;
        stringstream ss(s);

        while (ss >> word)
            words.push_back(word);

        if (pattern.length() != words.size())
            return false;

        unordered_map<char, string> map1;
        unordered_map<string, char> map2;

        for (int i = 0; i < pattern.length(); i++) {

            char ch = pattern[i];
            string w = words[i];

            if (map1.count(ch) && map1[ch] != w)
                return false;

            if (map2.count(w) && map2[w] != ch)
                return false;

            map1[ch] = w;
            map2[w] = ch;
        }

        return true;
    }
};

File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:

        words = s.split()

        if len(pattern) != len(words):
            return False

        map1 = {}
        map2 = {}

        for ch, word in zip(pattern, words):

            if ch in map1 and map1[ch] != word:
                return False

            if word in map2 and map2[word] != ch:
                return False

            map1[ch] = word
            map2[word] = ch

        return True